### ISSUE TITLE
Ability to keep animations unoptimized

### ISSUE THREAD

**Issue by Inviz** [zeux/meshoptimizer#185]
I'd like an ability to keep animation keyframes untouched by optimizations. 

My use case is: I create animation tracks in 3d editor, that later the game engine "reads" and turns into real character animations. For example I want to tell that for 2 seconds the character stands still, then walks left for 1 second, stands for 2, and jumps. That's 4 different animations, that I name in specific way that engine understands. I use `scale` property as a placeholder track. 

If i run `gltfpack` with no options, it says "There are no valid tracks". If i use `ac` option, if animation had only keyframes with the same values, it resets the tracks to have a single keyframe of always starting at `0` time. So the information on `start` and `end` time for the animation gets lost. It does not happen, if there's some actual movement.

So the solution in this case would be to allow animations to be untouched.

**Comment by Inviz** [comment]
<img width="972" alt="Screen Shot 2020-10-08 at 23 07 39" src="https://user-images.githubusercontent.com/2708/95477566-11f90980-09bb-11eb-9d9a-08f4d83def70.png">
Here's illustration of what i'm doing.

**Comment by zeux** [comment]
`-ac` was more or less intended for cases like this but you're right - it loses the information about the animation time range. If `-ac` emitted two keyframes for constant tracks, one at start time and one at end time, presumably that would work for your use case?

**Comment by Inviz** [comment]
Absolutely, that would work for me. Right now the start time is also set to 0, but in that case it should start where it has to. 

Thank you! Again, your library is amazing in what it does. I love how small and efficient it is, and it runs stupidly fast.

**Comment by zeux** [comment]
I looked into the existing code and I think I may have misunderstood the original problem. `-ac` indeed does discard the *end* time for the constant track, but it should preserve the starting time (which isn't sufficient for preserving the entire range of course).

Is it the case that for you, *all* animation tracks for the character that logically belong to different animations are part of the same animation in glTF, so there's effectively just one animation in glTF as opposed to multiple?

**Comment by Inviz** [comment]
In this case these are different animations. So ideally i'd get list of animations that only have 1 `scale` track, with (usually) 2 keyframes. They should not be merged (and they are not)

I do believe however, that if both keyframes are the same values, `-ac` discards starting time as well in my case. I'm getting a single keyframe with `0` as time, no matter when it starts. But if i make ending keyframe different (enough), suddenly starting time is correct, *and* there's a keyframe representing end time.

**Comment by Inviz** [comment]
Though I'll say that maybe I've used 0.14 version released a few months ago. I will try to confirm if it still happens on upstream version

**Comment by Inviz** [comment]
I think this is still happening on latest of gltpack. The animations that are "kept" by `-ac` get their keyframe reset to `0:00`

Here's input, output and blender source files attached. 

[keyframes.zip](https://github.com/zeux/meshoptimizer/files/5364936/keyframes.zip)



<img width="819" alt="Screen Shot 2020-10-12 at 19 50 32" src="https://user-images.githubusercontent.com/2708/95743336-33603b00-0cc4-11eb-8678-90d54ae803df.png">

**Comment by zeux** [comment]
Ah, yeah, apparently the resampling logic had a bug that caused all animations to start from 0 😓 Thanks for the report, I'm going to fix this and the issue with `-ac`.

**Comment by zeux** [comment]
Alright, so I've fixed the starting time, and I want to come back to the end time; I feel a bit hesitant about just outputting two key frames, as the current single-keyframe solution has important benefits - it minimizes the output file size, and it allows the runtime to trivially use a single-keyframe representation that doesn't require interpolation, which in some cases is valuable for performant implementations (as opposed to a two-keyframe track that requires comparing the output values).

Here's the status quo now with `-ac` setting:

- If the animation has any movement, the tracks with the movement are going to use the full time range, the tracks without the movement will just use a single starting keyframe placed at the original start of the animation
- If the animation has no movement, all tracks are going to be single-keyframe, losing information about the animation end

The question is, why is it important to know how long an animation is if the animation has no motion in it?

**Comment by Inviz** [comment]
Well for me it's very important, as i just want to mark up certain time ranges to dedicate them to play animations in my game engine. Without that i have to "fake" some movement, and hope it does not get quanitized away too. It just does not work in a meaningful way. I see where you are coming from about filesize, but I hope you can also see that timerange part is a very meaningful use case as well. Personally I can not conceptualize why is it gltfpack's responsibility tamper with animation keyframes. If somebody wanted 1-keyframe animations can't they do that in the editor? For that case, i think it's fine to output single time. However if the authors of animations made them play for a specific time (e.g. game engine may use this as a signal to stop rendering loop when time period is over), is it okay for gltfpack to lose that information?

I suspect you wouldn't want to complicate the api by introducing another cli option? Is there a way to re-use animation data for a different keyframe? I guess we can't have custom properties on animations either?

Right now I use a hack that I have to pre-process my files and write animation durations into their names, but it's a pretty harsh requirement that adds to complexity of pipeline (parsing and rewriting the files).

If you do think that's absolutely critical to make static animations single keyframe, i can keep my hack in place. But personally, i do not think it's a correct default behavior for a tool. 

Thanks!

**Comment by zeux** [comment]
On a fundamental level, gltfpack is designed to preserve the semantics of the glTF data, while being as free as possible to reinterpret structure. The core problem here is that glTF animations really don't have a concept of a time range (which is unfortunate! If this existed, we wouldn't have the problem in the first place).

This is a pretty central idea - for example, what happens when you say `-kn` is that gltfpack assumes that each individual node that is named can have its transform mutated at runtime, and - for example - blocks merging of meshes across different named nodes. When `-kn` is not specified, this constraint doesn't exist, and the optimization is unrestricted. But this should be understood as such, not as "kn disables mesh merging for meshes attached to named nodes" (it does, but that's not the problem statement - that's the solution).

So gltfpack doesn't "tamper" with animation keyframes in the sense that it aims to produce an animation that is semantically equivalent within a given visual tolerance; but it is free to rebuild the animation data in arbitrary ways when nothing constrains it to do so - this is more or less the point of the tool :)

Which then brings up a question as to whether the time range *implied* by the keyframe times is part of the definition of what an animation is in glTF or not.

If an animation *has* motion, then the range is pretty clearly important, because animations can loop.

If an animation does *not* have motion, then it's less clear - this is why I was asking the question. Is it the case that you have a sequencer that plays back individual tracks one after another, and because of this even if an animation doesn't have motion, it should be perceived as "stand still in this position for X seconds", and after X seconds pass another animation will be queued?

I'm asking these questions to have a precise understanding of the use case to make sure that the only reasonable solution is to double up the keyframes, not to suggest that I don't think something should be done here. Again in the ideal case glTF animations would have the time range information encoded explicitly in an optional manner, and gltfpack would preserve and/or reconstruct this information, but that's not the case unfortunately, so the core question is what the utility of a given part of glTF file is, and how to best preserve it.

**Comment by zeux** [comment]
Oh and one more question, for your use case does it matter that all tracks have the same time range, or does it only matter if the min & max of all of the tracks matches the original min & max?

**Comment by Inviz** [comment]
Thanks for thinking through it, zeux. I see your interpretation of what is fair game for gltfpack and I think it kinda makes sense. I still honestly don't exactly agree that removing information about when animation ends is within the semantics. But I concede that maybe I'm doing something weird. However the rest of ecosystem allows my concept to work, so I'd love if gltfpack also did allow it.

Yes, this is "stand still in position X", but also i play a different character clip based on the name of animation track. E.g. character nodding head, while physically not moving in space. For that I create an animation track from time S to time E, with the name "Character1 Nodding". Following the same logic, if animation track has some movement and is called "Walking", i'm playing one clip that physically moves character in space, and another that moves its legs to represent believable locomotion.

No, I do not want animation tracks to have the same time range. I want them to to have their own time ranges. That's how Blender exports them, and that suits my needs perfectly.  This is because I use those time in scheduling the animations. Some scenes have a bunch of overlapping tracks. I use additive animations (like nodding while walking) too.

**Comment by zeux** [comment]
@Inviz Right, what I meant was that in a given animation you might have different tracks with different ranges.

For example, in your input file from the comment above, there's an "Cafe/Character1/Walking" animation, and a "Cafe/Character2/Greeting/Upper/Additive" animation.

The walking animation has three tracks, for translation/rotation/scale; all three tracks have a time range (in the input file) of 0 .. 2.791667.
The greeting animation has one track, for scale; it has a time range (in the input file) of 1.5 .. 3.0.

After gltfpack processing (using master version), the output time ranges on the animations are as follows:

1. Cafe/Character1/Walking, translation & rotation tracks: time range preserved
2. Cafe/Character1/Walking, scale track: time range collapsed into a single point (0)
3. Cafe/Character2/Greeting/Upper/Additive, scale track: time range collapsed into a single point (1.5).

The fact that gltfpack collapses the greeting animation range (3) is a problem, as there's no way for you to recover the end of that range. However, is (2) also a problem, or is it fine because the time range of the entire animation is still recoverable from (1)?

**Comment by Inviz** [comment]
Ah I see what you mean here. I think that that I could handle that 2) case, and recover it from other tracks in animation by picking smallest and largest numbers. Even though it would be a little strange, i think it is not a big deal.

Thanks for helping me understand.

**Comment by zeux** [comment]
Alright, thanks. I'm going to look at the data in the test set I'm using for animations and come back with some sort of solution.

**Comment by zeux** [comment]
As documented in https://github.com/KhronosGroup/glTF/issues/1184 and a few other issues in glTF repository, a conformant implementation must do a union of all track's time ranges to compute the animation time range anyway (... there's also the bit about all animations starting from 0 apparently, which I am not sure I agree with but for the purposes of gltfpack that doesn't matter either way as it now preserves the start of the range).

Thus the problem only really exists for animations with no motion - for animations with motion, the time range is established by the tracks with motion inside the animation.

I evaluated three possible options to fix this:
- Duplicate keyframes in all constant tracks and use the time range from start to end
- Duplicate keyframes in all constant tracks but only for animations that have all tracks as constant
- Duplicate keyframes in a single constant track only for animations that have all tracks as constant

All of them solve the problem (as long as the loader computes time range as a union). The first option resulted in a measurable (although not critical) regression in size on a subset of the test models when `-ac` was used. The second and third options were roughly equivalent in the size impact (which was negligible).

Thus I decided on option 3 (which is slightly cleaner since as long as we need to preserve single-keyframe tracks for tracks with motion, we might as well do the absolute minimum amount of changes to tracks with no motion).
